### ISSUE TITLE
brew update feature: specify taps or only core

### ISSUE THREAD

**Issue by vladshablinsky** [Homebrew/homebrew#45196]
I noticed that `brew update` is pretty slow because it needs to get updates from the upstream. If you have a lot of taps, the update process can take a while. The more taps you have the more it takes `brew update` to update all the repositories. I think very often you want to update core and you don't wan't to wait until all the taps get updated. What if we have an option to get updates only from `Homebrew/homebrew` or update only specified taps? Something like `brew update --core` or `brew update caskroom/cask` ?

**Comment by dunn** [comment]
Sounds like a good idea to me.

**Comment by mikemcquaid** [comment]
I'd rather not add an option for this and instead approach ways to e.g. do the `git fetch`es in parallel.

**Comment by vladshablinsky** [comment]
@mikemcquaid thanks, maybe it's a better solution.

However, I don't know if I can implement it. I can come up with something like this or try to implement it with threads. 

```ruby
    tap_updaters = []
    Tap.each do |tap|
      next unless tap.git?
      tap_updaters << [Updater.new(tap.path), tap]
    end

    tap_updaters.each do |updater, tap|
      fork do
        updater.repository.cd do
          begin
            updater.pull!
          rescue Exception => e
            onoe "Failed to update tap: #{tap}"
          end
        end
      end
    end
    Process.waitall

    tap_updaters.each do |updater, tap|
      updated_taps << tap.name if updater.updated?
      report.update(updater.report) do |_key, oldval, newval|
        oldval.concat(newval)
      end
    end
```

**Comment by UniqMartin** [comment]
Personally, I think the biggest bottleneck is `git fetch` and not `git merge`/`git rebase`. Because there are very few things that can go wrong on a `git fetch`, I'd suggest (as a first step) to parallelize only that. This should be preceded by refactoring `Updater#pull!` into two separate methods.

Once we we have parallel fetch, I would suggest to look into running `Updater#report` in parallel, as that entails many invocations to `git` (to check if a version of a formula has changed, the old version needs to be extracted from the repository). Merging multiple reports into a single report is probably fast and not worth parallelizing.

Since I have no timings, the above is just what I think will give maximum benefit at very low risk. So the best first step would be to gather some timing data for the various steps of the update process.

@vladshablinsky If you want to work on this, don't miss out on what @xu-cheng already did in #40745.

**Comment by vladshablinsky** [comment]
@UniqMartin thanks for pointing that PR out. AFAIU the PR is exactly what @mikemcquaid told me about.

Agree with you on `Updater#report` stuff.

And closing this because there is a [PR](https://github.com/Homebrew/homebrew/pull/40745) pending.

**Comment by UniqMartin** [comment]
> And closing this because there is a PR pending.

Note that this PR hasn't been updated in quite a while.

If you are interested in working on this:

- Talk to @xu-cheng what his plans are for the PR. (He encouraged me recently to work on that, so I assume he hasn't the intention to pick it up himself soon.)
- Use this issue to flesh out how to approach this task. (I stated what I would do and in what order above, but it would be good to hear other opinions.)
- Once that's clear, create a tracking issue for the various steps and pursue them as separate PRs.

How does that sound?

**Comment by vladshablinsky** [comment]
@UniqMartin Initially I thought it would be just about adding a little `include?("--core")`  but now it becomes something more difficult. So, I can't work on it right now, but I would like to after I implement improvements for formula renames. However this feature can be implemented by the time I finish with formula renames. I'm going to start in several weeks, when I'm less busy. Now I can implement something easy and not time-consuming, therefore I can't work on this issue right now.

Thanks for all your suggestions though. If it's still relevant when I finish with formula renames, I'll be happy to work on it.

**Comment by mikemcquaid** [comment]
> Personally, I think the biggest bottleneck is git fetch and not git merge/git rebase. Because there are very few things that can go wrong on a git fetch, I'd suggest (as a first step) to parallelize only that. This should be preceded by refactoring Updater#pull! into two separate methods.

Agreed. I think this should be a dedicated PR itself as I'd bet money without looking at benchmarks by far the most time is spent in `git fetch` (other `git` operations and Ruby comparisons are pretty fast). 

I've been meaning to make a separate issue for this but I've been tempted for a while to actually rewrite part of `brew update` in shell-script so it's not dependent on the system Ruby being intact. The shell script could do the parallel `brew fetch` (or even `brew pull`) and store the last updated SHA-1 in either a file on disk or export it as an environment variable to the Ruby version. This could mean that the Ruby version of the `update` script is always already up-to-date with `master` before it gets run. This would make it much easier for us to fix updater bugs and help people who haven't updated in a while when e.g. OS X removes a system version of Ruby and also potentially migrate people to Tigerbrew/Linuxbrew when we deprecate Ruby 1.8 support.

> Thanks for all your suggestions though. If it's still relevant when I finish with formula renames, I'll be happy to work on it.

Thanks @vladshablinsky!
